src/TransformationUtils.cc

```diff
@@ -58,51 +58,92 @@ std::unique_ptr<TransitionSystem> toTransitionSystem(ChcDirectedGraph const & gr
         TermUtils utils(logic);
         if (isInit) {
             std::unordered_map<PTRef, PTRef, PTRefHash> subMap;
-            std::transform(edgeVars.nextStateVars.begin(), edgeVars.nextStateVars.end(), stateVars.begin(), std::inserter(subMap, subMap.end()),
+            std::transform(edgeVars.nextStateVars.begin(), edgeVars.nextStateVars.end(), stateVars.begin(),
+                           std::inserter(subMap, subMap.end()),
                            [](PTRef key, PTRef value) { return std::make_pair(key, value); });
             init = utils.varSubstitute(fla, subMap);
             init = QuantifierElimination(logic).keepOnly(init, stateVars);
-//            std::cout << logic.printTerm(init) << std::endl;
+            //            std::cout << logic.printTerm(init) << std::endl;
         }
         if (isLoop) {
             transitionRelation = transitionFormulaInSystemType(*systemType, edgeVars, fla, logic);
-//            std::cout << logic.printTerm(transitionRelation) << std::endl;
+            //            std::cout << logic.printTerm(transitionRelation) << std::endl;
         }
         if (isEnd) {
             std::unordered_map<PTRef, PTRef, PTRefHash> subMap;
-            std::transform(edgeVars.stateVars.begin(), edgeVars.stateVars.end(), stateVars.begin(), std::inserter(subMap, subMap.end()),
+            std::transform(edgeVars.stateVars.begin(), edgeVars.stateVars.end(), stateVars.begin(),
+                           std::inserter(subMap, subMap.end()),
                            [](PTRef key, PTRef value) { return std::make_pair(key, value); });
             bad = utils.varSubstitute(fla, subMap);
             bad = QuantifierElimination(logic).keepOnly(bad, stateVars);
-//            std::cout << logic.printTerm(bad) << std::endl;
+            //            std::cout << logic.printTerm(bad) << std::endl;
         }
     });
     assert(init != PTRef_Undef && transitionRelation != PTRef_Undef && bad != PTRef_Undef);
-    auto ts = std::unique_ptr<TransitionSystem>(new TransitionSystem(logic, std::move(systemType), init, transitionRelation, bad));
+    auto ts = std::unique_ptr<TransitionSystem>(
+        new TransitionSystem(logic, std::move(systemType), init, transitionRelation, bad));
     return ts;
 }
 
-bool isTransitionSystemChain(ChcDirectedGraph const & graph) {
+bool strongConnection(std::set<int> & visitedVertices, std::set<int> & verticesOnStack,
+                      AdjacencyListsGraphRepresentation & graphRepresentation, ChcDirectedGraph const & graph,
+                      SymRef node) {
+    visitedVertices.insert(node.x);
+    verticesOnStack.insert(node.x);
+    auto const & outEdges = graphRepresentation.getOutgoingEdgesFor(node);
+
+    for (EId eid : outEdges) {
+        if (size(outEdges) <= 1) { return false; }
```

Comment:
Is there an implicit assumption that if there is exactly one out-edge, it is a self-loop?

Reply:
Actually, why is this inside the loop?

Reply:
Yup, I basically expect that if there's just one outgoing edge, then there is no loop for this specific node, bc it means that we just have a self-loop here.
I'll move it outside the loop

Reply:
Fixed